packages/eslint-plugin/tests/rules/consistent-generic-constructors.test.ts

```diff
@@ -0,0 +1,151 @@
+import rule from '../../src/rules/consistent-generic-constructors';
+import { RuleTester, noFormat } from '../RuleTester';
+
+const ruleTester = new RuleTester({
+  parser: '@typescript-eslint/parser',
+});
+
+ruleTester.run('consistent-generic-constructors', rule, {
+  valid: [
+    // default: rhs
+    'const a = new Foo();',
+    'const a = new Foo<string>();',
+    'const a: Foo<string> = new Foo<string>();',
+    'const a: Foo = new Foo();',
+    'const a: Bar<string> = new Foo();',
+    'const a: Foo = new Foo<string>();',
+    'const a: Bar = new Foo<string>();',
+    'const a: Bar<string> = new Foo<string>();',
+    'const a: Foo<string> = Foo<string>();',
+    'const a: Foo<string> = Foo();',
+    'const a: Foo = Foo<string>();',
+    // lhs
+    {
+      code: 'const a = new Foo();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a: Foo<string> = new Foo();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a: Foo<string> = new Foo<string>();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a: Foo = new Foo();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a: Bar = new Foo<string>();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a: Bar<string> = new Foo<string>();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a: Foo<string> = Foo<string>();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a: Foo<string> = Foo();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a: Foo = Foo<string>();',
+      options: ['lhs'],
+    },
+    {
+      code: 'const a = new (class C<T> {})<string>();',
```

Comment:
Great test case 😄 I don't know that I would have thought of it.

Trying to think creatively, a couple more valid cases to add...

* `const a: Foo<number> = new Foo<0 | 1>`
* `const a: Foo<0 | 1> = new Foo<number>`

Also, just to be safe, how about some `invalid` cases around including comments and/or multiline?

* `const x: Foo /* ... */ <string> = new Foo()`
* `const x: Foo /* ... */ <string> = new Foo /* ... */ <string> /* .. */ ()`
* `const x: Foo<string> = new \n Foo()`
* `const x: Foo = new \n Foo<string>()`

Reply:
> `const a: Foo<number> = new Foo<0 | 1>`

That's not valid code, but that inspired me. My fixer doesn't work if the RHS doesn't contain parens.

Reply:
> That's not valid code

[It is actually](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAHgLhgZQKZQDxgK4FsBGqATgHwwC8MYqA7iuhgAwwA+MAjCQBQCUA3EA)!! TS is very unsafe with its generics and doesn't enforce variance like you'd like from a strictly typed language.

Fun fact - [flow does actually enforce the variance correctly here](https://flow.org/try/#0MYewdgzgLgBAHgLhgZQKZQDxgK4FsBGqATgHwwC8MYqA7iuhgAwwA+MAjCQBQCUA3EA)

Reply:
I said because I tried :(

[Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEkoGdnwGIHtMB4AqAfANwCwAUOWJgHbIAu8UAXBtjtQK4C2ARiDAXgBeeNRAB3VrgAM8AD7wAjAXJA)

(Without the parens)

Reply:
Fun fact - [if you manually control the variance of the generic using 4.7's features](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220513#code/CYUwxgNghgTiAEkoGdnwGIHtMB4CWAdvACoB88A3gL4CwAUPWJgcgC7xQBcG2OBArgFsARiBjkAvPAIgA7j1wAGeAB94ARlIAKAJQBuekA) - then you can make this error as it's supposed to.

Reply:
> I said because I tried :(

Funnily you tried in v4.6... in v4.7 `new Foo<0 | 1>` is now valid thanks to the "expression with type arguments" feature!